electionleaflets/apps/leaflets/forms.py

```diff
@@ -125,20 +129,23 @@ def get_ballot_data_from_ynr(self, postcode):
 
         :type instance: Leaflet
         """
-        params = {"for_postcode": postcode}
-        start, end = self.get_date_range()
-        params["election_date_range_after"] = start
-        params["election_date_range_before"] = end
-        ynr_helper = YNRAPIHelper()
-        try:
-            resp = ynr_helper.get(
-                "ballots",
-                params=params,
-                json=True,
-            )
-        except requests.RequestException:
-            return []
-        return resp["results"]
+        if "ynr_data" not in self.storage.extra_data:
+            params = {"for_postcode": postcode}
+            start, end = self.get_date_range()
+            params["election_date_range_after"] = start
+            params["election_date_range_before"] = end
+            ynr_helper = YNRAPIHelper()
+            try:
+                resp = ynr_helper.get(
+                    "ballots",
+                    params=params,
+                    json=True,
+                )
+            except requests.RequestException:
+                return []
+
+            self.storage.extra_data["ynr_data"] = resp["results"]
+        return self.storage.extra_data["ynr_data"]
```

Comment:
In the constructor, you're assigning

```py
self.storage = kwargs.get("initial", {}).get("storage", None)
```

so that means the value of `self.storage` can be `None` here.
But we're accessing `self.storage.extra_data`, which could raise

```
AttributeError: 'NoneType' object has no attribute 'extra_data'
```

Looks like we need to make sure it is safe to access this attribute first

Reply:
I've switched to using get_form_kwargs. This doesn't in itself solve the problem, but it does mean we don't need to rely on the structure of the initial kwarg.

I've also made it blow up if not provided, so we'll catch any problem earlier.